### PR TITLE
Scahin/automate infra proxy ui

### PR DESCRIPTION
**Pull request by sachin-msys**
### :nut_and_bolt: Description: What code changed, and why?
Code changes for introduction of `infra proxy module` in `automate-ui` In order to make a single source of truth for chef infra servers data here

**In this PR following points we have covered:**

1. Created `infra-proxy` module in automate-ui to keep all `infra-proxy` related component separate
2.  Designed `infra proxy`  UI flows with mock data

**Pending Task**
1. `infra-proxy` API integration with UI

### :chains: Related Resources
Refered Wireframes: https://chef.invisionapp.com/share/PZSLN5TJ8B9#/screens/369724917
For feature, description refer: #1544 
### :+1: Definition of Done
I have added infra proxy module and UI new pages
1. chef server's list page
2. create chef-server popup modal
3. chef-server details page 
4. org create and edit popup modal 
### :athletic_shoe: How to Build 
build components/automate-ui

### CODE REVIEW

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                <th class="id-column">IP Address</th>
+              </tr>
+            </thead>
+            <tbody>
+              <tr class="detail-row">
+                <td class="id-column">chef.io</td>
+                <td class="id-column">192.168.1.1</td>
+              </tr>
+            </tbody>
+          </table>
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html]: we have a new angular component named `chef-table-new` that you might be able to use here.  It's under `app/components/table/table.component.ts`

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.scss [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+form {
+
+  #update-name {
+    width: 240px;
+  }
+
+  input {
+    font-size: 1em;
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.scss]: FYI for in the near future, `font-size` on `chefInput` directives _shouldn't_ be needed, and will be handled by the chefInput directive at 14px. 😄

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        (close)="closeCreateModal()"
+        [conflictErrorEvent]="conflictErrorEvent"
+        (createClicked)="createChefServer()">
+      </app-create-chef-server-modal>
+
+      <div class="page-body">
+        <chef-toolbar>
+          <chef-button primary (click)="openCreateModal()">Add Chef Server</chef-button>
+        </chef-toolbar>
+        <chef-table>
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html]: same as above - we now have `chef-table-new` and its corresponding table elements in non-stencil form

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      <chef-tab-selector [value]="tabValue" (change)="onSelectedTab($event)">
+        <chef-option value='orgs' data-cy="orgs-tab">Orgs</chef-option>
+        <chef-option value='details' data-cy="details-tab">Details</chef-option>
+      </chef-tab-selector>
+    </chef-page-header>
+    <section class="page-body" *ngIf="tabValue === 'details'">
+      <form [formGroup]="chefServerForm">
+        <chef-form-field id="update-name">
+            <label>
+                <span class="label">Disaplay Name <span aria-hidden="true">*</span></span>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html]:
```suggestion
                <span class="label">Chef Server Name <span aria-hidden="true">*</span></span>
```

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            <chef-tbody>
+              <chef-tr>
+                <chef-td>
+                  <a (click)="openCreateModal('edit')">Automate Team</a>
+                </chef-td>
+                <chef-td>andhfngth</chef-td>
+                <chef-td>Inherited from chef server</chef-td>
+                <chef-td class="controls">
+                  <chef-control-menu>
+                      <chef-option  class="delete-rule" data-cy="delete-rule">Delete Orgs</chef-option>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html]:
```suggestion
                      <chef-option  class="delete-rule" data-cy="delete-rule">Delete Org</chef-option>
```

File: components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html [hunk 6]
```diff
@@ -0,0 +1,33 @@
+<chef-modal [visible]="visible" (closeModal)="closeEvent()">
+  <h2 *ngIf="type === 'create'" slot="title">Create New Org</h2>
+  <h2 *ngIf="type === 'edit'" slot="title">Create New Org</h2>
+  <div class="flex-container">
+    <form [formGroup]="orgForm">
+      <div class="name-margin">
+        <chef-form-field>
+          <label>
+            <span class="label">Full Name <span aria-hidden="true">*</span></span>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html]:
```suggestion
            <span class="label">Org Name <span aria-hidden="true">*</span></span>
```

File: components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            <input chefInput formControlName="name" type="text" id="name-input" data-cy="full-name" autocomplete="off">
+          </label>
+          <chef-error *ngIf="orgForm.get('name').dirty">
+            Name is required.
+          </chef-error>
+        </chef-form-field>
+        <chef-from-field>
+          <label>
+            <span class="label">Org Id <span aria-hidden="true">*</span></span>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html]:
```suggestion
            <span class="label">Org ID <span aria-hidden="true">*</span></span>
```

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+                  <chef-td>
+                    <a [routerLink]="['/infrastructure/chef-servers', server.name]">{{ server.name }}</a>
+                  </chef-td>
+                  <chef-td>{{ server.fqdn }}</chef-td>
+                  <chef-td>{{ server.ip_address }}</chef-td>
+                  <chef-td>{{ server.orgs }}</chef-td>
+                  <chef-td class="controls">
+                    <app-authorized>
+                      <chef-control-menu>
+                        <chef-option>Remove Server</chef-option>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html]:
```suggestion
                        <chef-option>Remove Chef Server</chef-option>
```

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html [hunk 9]
```diff
@@ -0,0 +1,51 @@
+<app-client-runs-sidebar></app-client-runs-sidebar>
+
+<div class="container">
+    <main>
+      <chef-page-header>
+        <chef-heading>Chef Servers</chef-heading>
+        <chef-subheading>Manage Chef Servers with Chef Automate</chef-subheading>
```
Comment (susanev) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html]:
```suggestion
        <chef-subheading>Manage Chef Servers with Chef Automate.</chef-subheading>
```

**Comment by jonong1972** [comment]
I noticed that "Chef Servers" link is on under the "Infrastructure" section.  Was there a reason why it was moved here instead of under the "Settings" section?  In the mock ups:

https://chef.invisionapp.com/share/PZSLN5TJ8B9#/screens/369724917

It would exist under the Node Management group  above "Node Integrations"

I just noticed that in the top navigation in my mockups are missing the Settings section. My fault for that.

I heard that our UI team is currently working on the Settings side nav as well. You may want to reach out to @Shadae about this?

**Comment by jonong1972** [comment]
Is it possible to have the "Create" modal have a dynamic modal title? Currently when I click on the Create Og, the modal title says "Create New Org"

When I click the link for "Automate Team"  the modal title still says "Create New Org".  I believe I had that as "Edit Org" in the mock up
https://chef.invisionapp.com/share/PZSLN5TJ8B9#/screens/369728100

**Comment by susanev** [comment]
@jonong1972 the editing will happen from the details tab not the modal right? the modal is only for creating i believe.

**Comment by sachin-msys** [comment]
> I noticed that "Chef Servers" link is on under the "Infrastructure" section. Was there a reason why it was moved here instead of under the "Settings" section? In the mock ups:

As per the team discussion with @btm ` chef-servers ` is related to ` Infrastructure` so it was suggested that to move it under ` Infrastructure Tab `

**Comment by Shadae** [comment]
Hi Sachin - double checking you still want this PR in draft mode 😄

**Comment by sachin-msys** [comment]
@Shadae - Hey as we are working on new "infra-proxy-service" so we want to keep this PR in progress until we finalise the stable state.
Ref: https://github.com/chef/automate/issues/1544

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+          [visible]="deleteModalVisible"
+          objectNoun="server"
+          [objectName]="serverToDelete?.name"
+          (close)="closeDeleteModal()"
+          (deleteClicked)="deleteServer()"
+          objectAction="Delete">
+        </app-delete-object-modal>
+        <div class="page-body">
+          <chef-toolbar>
+            <chef-button primary (click)="openCreateModal()">Add Chef Server</chef-button>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html]: s/Chef Server/Chef Infra Server?

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+          </chef-toolbar>
+          <chef-table-new>
+            <chef-table-header>
+              <chef-table-row>
+                <chef-table-header-cell>Chef Server Name</chef-table-header-cell>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html]: - Replace `Chef Server Name` with `Name`.
- The Description should be a placeholder of Name.
- The Name should be treated as an ID(unique).

Ref: Team/Project Create Model popup.
project-create-model

File: components/automate-ui/src/app/pages/infra-proxy/create-chef-server-modal/create-chef-server-modal.component.html [hunk 12]
```diff
@@ -0,0 +1,47 @@
+<chef-modal [visible]="visible" (closeModal)="closeEvent()">
+  <h2 slot="title">Add Chef Server </h2>
+    <div class="flex-container">
+      <form [formGroup]="createForm">
+        <div class="margin">
+          <chef-form-field>
+            <label>
+              <span class="label">Name <span aria-hidden="true">*</span></span>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/create-chef-server-modal/create-chef-server-modal.component.html]: - The Description should be a placeholder of Name.
- The Name should be treated as an ID(unique).

Ref: Team/Project Create Model popup.
project-create-model

File: components/automate-ui/src/app/pages/infra-proxy/create-chef-server-modal/create-chef-server-modal.component.html [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            <label>
+              <span class="label">IP Address <span aria-hidden="true">*</span></span>
+              <input chefInput formControlName="ip_address" type="text" id="ip-address-input" autocomplete="off"/>
+            </label>
+          </chef-form-field>
+        </div>
+        <div id="button-bar">
+          <chef-button primary id="create-button-object-modal" (click)="createChefServer()">
+            <chef-loading-spinner *ngIf="creating"></chef-loading-spinner>
+            <span *ngIf="!creating">Add Chef Server</span>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/create-chef-server-modal/create-chef-server-modal.component.html]: s/Chef Server/Chef Infra Server?

Applicable for all places

File: components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        </chef-form-field>
+      </div>
+      <div id="button-bar">
+        <chef-button primary id="create-org-modal-btn" (click)="createServerOrg()">Create Org</chef-button>
+        <chef-loading-spinner *ngIf="creating"></chef-loading-spinner>
+        <chef-button tertiary (click)="closeEvent()">Cancel</chef-button>
+      </div>
+    </form>
+  </div>
+</chef-modal>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html]: [minor] Add newline at the end

File: components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+            <span class="label">Admin User <span aria-hidden="true">*</span></span>
+            <input chefInput formControlName="admin_user" type="text" id="admin-input" data-cy="admin-user" autocomplete="off">
+          </label>
+        </chef-form-field>
+        <chef-form-field>
+          <label>
+            <span class="label">Admin Key <span aria-hidden="true">*</span></span>
+            <textarea name="admin_key" id="admin_key" formControlName="admin_key" cols="57" rows="10" autofocus></textarea>
+          </label>
+        </chef-form-field>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html]: Allow all users access to Infra Views option
 - On the orgs settings page, add a toggle to allow all authenticated users access to the API or just admins. Would be greyed out for non-admins. Defaults to admins only.

File: components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html [hunk 16]
```diff
@@ -0,0 +1,32 @@
+<chef-modal [visible]="visible" (closeModal)="closeEvent()">
+  <h2 slot="title">Create New Org</h2>
+  <div class="flex-container">
+    <form [formGroup]="createForm">
+      <div class="name-margin">
+        <chef-form-field>
+          <label>
+            <span class="label">Org Name <span aria-hidden="true">*</span></span>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html]: [minor] s/Org Name/Name

File: components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+        <chef-form-field>
+          <label>
+            <span class="label">Admin User <span aria-hidden="true">*</span></span>
+            <input chefInput formControlName="admin_user" type="text" id="admin-input" data-cy="admin-user" autocomplete="off">
+          </label>
+        </chef-form-field>
+        <chef-form-field>
+          <label>
+            <span class="label">Admin Key <span aria-hidden="true">*</span></span>
+            <textarea name="admin_key" id="admin_key" formControlName="admin_key" cols="57" rows="10" autofocus></textarea>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/org-create-edit-modal/org-create-edit-modal.component.html]: [minor] Add placeholder `-----BEGIN RSA PRIVATE KEY-----`

Ref: 
textarea

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+      </section>
+      <section class="page-body" *ngIf="tabValue === 'orgs'">
+        <ng-container>
+          <chef-toolbar>
+            <chef-button id="create-button" primary (click)="openCreateModal('create')">Create Org</chef-button>
+          </chef-toolbar>
+          <chef-table-new>
+            <chef-table-header>
+              <chef-table-row>
+                <chef-table-header-cell>Org Name</chef-table-header-cell>
```
Comment (vsingh-msys) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html]: [nit]s/Org Name/Name

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.ts [hunk 19]
```diff
@@ -182,4 +188,33 @@ export class ChefServersDetailsComponent implements OnInit, OnDestroy {
...
+    this.saving = true;
+    const updatedServer = {
+      id: this.server.id,
+      name: this.updateServerForm.controls.name.value.trim(),
+      description: this.updateServerForm.controls.description.value.trim(),
+      fqdn: this.updateServerForm.controls.fqdn.value.trim(),
+      ip_address: this.updateServerForm.controls.ip_address.value.trim()
+    };
+    this.store.dispatch(new UpdateServer({server: updatedServer}));
+    const pendingSave = new Subject<boolean>();
```
Comment (msorens) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.ts]: We have recently identified this `pendingSave` (or `pendingAdd`, etc.) as an anti-pattern. You can find several PRs from about a month ago that have replaced this with a much cleaner, simpler approach.
Example PRs:
https://github.com/chef/automate/pull/2188
https://github.com/chef/automate/pull/2205
https://github.com/chef/automate/pull/2206

Please update accordingly.

File: components/automate-ui/src/app/entities/servers/server.reducer.ts [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+  saveStatus: EntityStatus;
+  saveError: HttpErrorResponse;
+  getStatus: EntityStatus;
+  updateStatus: EntityStatus;
+  deleteStatus: EntityStatus;
+}
+
+// reusable names of the above properties:
+// TODO (tc): Rename this to GET_ALL_STATUS
+const STATUS = 'status';
```
Comment (msorens) [inline components/automate-ui/src/app/entities/servers/server.reducer.ts]: Please do what the comment says--rename to `GET_ALL_STATUS` (and `getAllStatus`). Prefer not to propagate a name that is less useful.

File: components/automate-ui/src/app/pages/infra-proxy/cookbook-list/cookbooks-list.component.ts [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  saveOrg(): void {
+    this.saveSuccessful = false;
+    this.saving = true;
+    this.store.dispatch(new UpdateOrg({
+      org: this.org
+    }));
+
+    const pendingSave = new Subject<boolean>();
```
Comment (msorens) [inline components/automate-ui/src/app/pages/infra-proxy/cookbook-list/cookbooks-list.component.ts]: Same comment as above about removing the `pendingSave` construct here.

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+                <chef-table-header-cell class="controls"></chef-table-header-cell>
+              </chef-table-row>
+            </chef-table-header>
+            <chef-table-body>
+              <chef-table-row *ngFor="let org of orgs">
+                <chef-table-cell>
+                  <a [routerLink]="['/infrastructure','chef-servers', server?.id, 'orgs', org.id, 'cookbooks']">{{ org.name }}</a>
+                </chef-table-cell>
+                <chef-table-cell>{{ org.admin_user }}</chef-table-cell>
+                <chef-table-cell class="controls">
```
Comment (msorens) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-details/chef-servers-details.component.html]: Another recent change was to standardize on "three-dot-column" for this class wrapping the control menu.
```suggestion
                <chef-table-cell class="three-dot-column">
```

File: components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+          </chef-table-header>
+          <chef-table-body>
+            <chef-table-row *ngFor= "let server of sortedChefServers$ | async">
+              <chef-table-cell>
+                <a [routerLink]="['/infrastructure/chef-servers', server.id]">{{ server.name }}</a>
+              </chef-table-cell>
+              <chef-table-cell>{{ server.fqdn }}</chef-table-cell>
+              <chef-table-cell>{{ server.ip_address }}</chef-table-cell>
+              <chef-table-cell>{{ server.description }}</chef-table-cell>
+              <chef-table-cell class="controls">
```
Comment (msorens) [inline components/automate-ui/src/app/pages/infra-proxy/chef-servers-list/chef-servers-list.component.html]: Same comment as above--change to standardized "three-dot-column".

**Comment by btm** [comment]
> > I noticed that "Chef Servers" link is on under the "Infrastructure" section. Was there a reason why it was moved here instead of under the "Settings" section? In the mock ups:
> 
> As per the team discussion with @btm `chef-servers` is related to ` Infrastructure` so it was suggested that to move it under `Infrastructure Tab`

We should have the work that's done so far on the Settings tab. That is, adding the Chef Server and Organizations is something you'll rarely do so we want to move that under Settings.

When we get to the views like showing the cookbooks, roles, and other artifacts, we want that under Chef Servers on the Infrastructure tab (for now).

**Comment by btm** [comment]
The 'Create Org' button should be renamed to 'Add Organization' to match 'Add Chef Server' (and we're not creating an org on the chef server, we're adding credentials to access one. 

We might reconsider this language to something like 'Link Chef Server' and 'Link Organization' so the user understands they're not modifying anything on the Chef Server here, just connecting it to Automate. @chef/ux-team  ?

**Comment by susanev** [comment]
> The 'Create Org' button should be renamed to 'Add Organization' to match 'Add Chef Server' (and we're not creating an org on the chef server, we're adding credentials to access one.
> 
> We might reconsider this language to something like 'Link Chef Server' and 'Link Organization' so the user understands they're not modifying anything on the Chef Server here, just connecting it to Automate. @chef/ux-team ?

@btm _connect_ might be a better verb then _link_ though since _link_ has a very specific meaning on the web.

changing _Create Org_ to _Add Organization_ sounds good to me, we should make sure to update the modal text too in addition to the button text.

**Comment by btm** [comment]
@sachin-msys did #2581 or another PR replace this? can this be closed?

**Comment by sachin-msys** [comment]
> @sachin-msys did #2581 or another PR replace this? can this be closed?

@btm we have broken this into three smaller one 
1. #2581 for infra server module
2. #2784 for infra org module
3. #3014 for infra cookbook list module 
this one is no longer needed so this can be closed

**Comment by btm** [comment]
looks like this draft PR was split into multiple smaller PRs:
server list: #2581
server detail/orgs: #2784
cookbooks list: #3014

**Pull request closed without merging**